runtime/script_loader.cpp

```diff
@@ -167,15 +272,59 @@ JSObject* module_resolve_hook(JSContext* cx, HandleValue referencingPrivate,
   return get_module(cx, path.get(), resolved_path, opts);
 }
 
- ScriptLoader::ScriptLoader(JSContext *cx, JS::CompileOptions *opts) {
+bool module_metadata_hook(JSContext* cx, HandleValue referencingPrivate, HandleObject metaObject) {
+  RootedObject info(cx, &referencingPrivate.toObject());
+  RootedValue parent_id_val(cx);
+  if (!JS_GetProperty(cx, info, "id", &parent_id_val)) {
+    return false;
+  }
+  if (!parent_id_val.isString()) {
+    return false;
+  }
+  RootedValue builtin_val(cx);
+  if (!MapGet(cx, builtinImports, parent_id_val, &builtin_val)) {
+    return false;
+  }
+  if (builtin_val.isUndefined()) {
+    return false;
+  }
+  JS_SetProperty(cx, metaObject, "builtin", builtin_val);
+  return true;
+}
+
+ScriptLoader::ScriptLoader(JSContext *cx, JS::CompileOptions *opts) {
   MOZ_ASSERT(!SCRIPT_LOADER);
   SCRIPT_LOADER = this;
   CONTEXT = cx;
   COMPILE_OPTS = opts;
   moduleRegistry.init(cx, JS::NewMapObject(cx));
+  builtinImports.init(cx, JS::NewMapObject(cx));
```

Comment:
I guess in principle we could keep to just having the ` moduleRegistry` map, and initializing the entries for uninitialized builtin modules to some magic value, which is then detected in the resolution hook and replaced with the initialized module.

That doesn't have to block landing this though, as I think we can live with the current approach perfectly well :)

Reply:
Yes, having a module registry with structured data I think would make sense as soon as we need more context in future. Possibly even an unordered map to structs with rooting.

Reply:
For now I'll leave this as-is though.